core/deliverservice/bft_client.go

```diff
@@ -0,0 +1,539 @@
+	msgCryptoVerifier  MessageCryptoVerifier     // verifies headers
+
+	// The total time a bft client tries to connect (to all available endpoints) before giving up leadership
+	reconnectTotalTimeThreshold time.Duration
+	// the total time a block receiver tries to connect before giving up and letting the client to try another
+	// endpoint as a block receiver
+	reconnectBlockRcvTotalTimeThreshold time.Duration
+	// The minimal time between connection retries. This interval grows exponentially until maxBackoffDelay, below.
+	minBackoffDelay time.Duration
+	// The maximal time between connection retries.
+	maxBackoffDelay time.Duration
+	// The block censorship timeout. A block censorship suspicion is declared if more than f header receivers are
+	// ahead of the block receiver for a period larger than this timeout.
+	blockCensorshipTimeout time.Duration
+
+	endpoints          []comm.EndpointCriteria // a set of endpoints
+	blockReceiverIndex int                     // index of the current block receiver endpoint
+
+	blockReceiver   *broadcastClient
+	nextBlockNumber uint64
+	lastBlockTime   time.Time
+
+	headerReceivers map[string]*bftHeaderReceiver
+}
+
+func NewBFTDeliveryClient(
+	chainID string,
+	connFactory comm.ConnectionFactory,
+	endpoints []comm.EndpointCriteria,
+	clFactory clientFactory,
+	ledgerInfoProvider blocksprovider.LedgerInfo,
+	msgVerifier MessageCryptoVerifier,
+) *bftDeliveryClient {
+	c := &bftDeliveryClient{
+		stopChan:                            make(chan struct{}, 1),
+		chainID:                             chainID,
+		connectionFactory:                   connFactory,
+		createClient:                        clFactory,
+		ledgerInfoProvider:                  ledgerInfoProvider,
+		msgCryptoVerifier:                   msgVerifier,
+		reconnectTotalTimeThreshold:         getReConnectTotalTimeThreshold(),
+		reconnectBlockRcvTotalTimeThreshold: util.GetDurationOrDefault("peer.deliveryclient.bft.blockRcvTotalBackoffDelay", bftBlockRcvTotalBackoffDelay),
+		minBackoffDelay:                     util.GetDurationOrDefault("peer.deliveryclient.bft.minBackoffDelay", bftMinBackoffDelay),
+		maxBackoffDelay:                     util.GetDurationOrDefault("peer.deliveryclient.bft.maxBackoffDelay", bftMaxBackoffDelay),
+		blockCensorshipTimeout:              util.GetDurationOrDefault("peer.deliveryclient.bft.blockCensorshipTimeout", bftBlockCensorshipTimeout),
+		endpoints:                           comm.Shuffle(endpoints),
+		blockReceiverIndex:                  -1,
+		headerReceivers:                     make(map[string]*bftHeaderReceiver),
+	}
+
+	bftLogger.Infof("[%s] Created BFT Delivery Client", chainID)
+	return c
+}
+
+func (c *bftDeliveryClient) Recv() (response *orderer.DeliverResponse, err error) {
+	bftLogger.Debugf("[%s] Entry", c.chainID)
+
+	c.startOnce.Do(func() {
+		var num uint64
+		num, err = c.ledgerInfoProvider.LedgerHeight()
+		if err != nil {
+			return
+		}
+		c.nextBlockNumber = num
```

Comment:
this is inside the once closure, but what if we get a block via gossip and we commit it?